File: java8-scala/src/java8/Main02.scala
package java8

/** An example of traits. */
object Main02 {
  def main(args: Array[String]) {
    val f = new FooImpl()
    f.sayHi()
    f.sayBye()
  }

  trait Foo {
    def sayHi()
    
    def sayBye() = println("bye")
  }
  
  trait Zaz {
    // def sayBye() = println("bye!")
  }
  
  class FooImpl extends Foo with Zaz {
    override def sayHi() {
      println("hi");
    }
  }
}

File: java8/src/java8/Main02.java
package java8;

/** Example of multiple default methods. */
public class Main02 {
  public static void main(String[] args) {
    Foo f = new FooImpl();
    f.sayHi();
    f.sayBye();
  }

  public interface Foo {
    void sayHi();

    default void sayBye() {
      System.out.println("bye");
    }
  }

  public interface Zaz {
    default void sayBye() {
      System.out.println("bye!");
    }
  }

  public static class FooImpl implements Foo, Zaz {
    @Override
    public void sayHi() {
      System.out.println("hi");
    }

    @Override
    public void sayBye() {
      Zaz.super.sayBye();
    }
  }
}
